File: src/Solutions/Euler622.hs
module Solutions.Euler622 where
import Utils.NumberTheory(divisors)

-- http://oeis.org/A002326

runShuffleEq :: Int -> Int -> Bool
runShuffleEq bound n = go (bound - 1) 2
    where go left l | left <= 0 = l == 1
                    | l == 1 = False
                    | otherwise = go (left - 1) ((2 * l) `mod` n)

sumS :: Int -> Int
sumS n = sum $ map succ $ filter (runShuffleEq n) $ tail $ divisors (2^n - 1)

euler622 :: IO String
euler622 = return $ show $ sumS 60

File: rust/src/solutions/euler622.rs
use crate::utils::numeric::divisors::divisors;

fn run_shuffle_eq(bound: i64, n: i64) -> bool {
    let mut left = bound - 1;
    let mut l = 2;
    loop {
        if left <= 0 {
            break l == 1;
        } else if l == 1 {
            break false;
        } else {
            left -= 1;
            l = (l + l) % n;
        }
    }
}

fn sum_s(n: u32) -> i64 {
    divisors(2i64.pow(n) - 1)
        .into_iter()
        .skip(1)
        .filter(|&l| run_shuffle_eq(n as i64, l))
        .map(|l| l + 1)
        .sum()
}

pub fn euler622() -> String {
    sum_s(60).to_string()
}
